src/host/getset.cpp

```diff
@@ -1028,36 +1036,89 @@ void ApiRoutines::GetLargestConsoleWindowSizeImpl(const SCREEN_INFORMATION& cont
                 return S_OK;
             }
 
-            const auto clipViewport = clip ? Viewport::FromInclusive(*clip) : bufferSize;
+            const auto clipViewport = clip ? Viewport::FromInclusive(*clip).Clamp(bufferSize) : bufferSize;
             const auto sourceViewport = Viewport::FromInclusive(source);
-            Viewport readViewport;
-            Viewport writtenViewport;
+            const auto fillViewport = sourceViewport.Clamp(clipViewport);
 
-            const auto w = std::max(0, sourceViewport.Width());
-            const auto h = std::max(0, sourceViewport.Height());
-            const auto a = static_cast<size_t>(w * h);
-            if (a == 0)
+            writer.BackupCursor();
+
+            if (gci.GetVtIo()->GetDeviceAttributes().test(Microsoft::Console::VirtualTerminal::DeviceAttribute::RectangularAreaOperations))
             {
-                return S_OK;
+                // This calculates just the positive offsets caused by out-of-bounds source and target coordinates.
+                // The source gets offset because it's outside the bounds of the buffer (= negative),

```

Comment:
this might benefit from some ASCII art. I still don't understand it

Reply:
If the source starts outside the buffer, we want to offset it by +source. if the target starts outside the clip region, we want to offset something by something.

Reply:
I'm surprised we can apply the same offset to both the source and target!

Reply:
I believe there's no good way to draw this. It's a logic that I came up with after noodling over the problem for a while.